File: 0050-powx-n/0050-powx-n.cpp
class Solution {
public:
    
    // unordered_map<int,double>gzip;
    double recursion(double x,int n){
        
        
               
                if(n==0){
                    
                        return 1;
                }
                
                if(n==1){
                        return x;
                }
            
        
                double val;
                if(n%2==0){
                    
                        val=recursion(x,n/2);
                        return val*val;
                }    
                
                else{
                    
                        val=recursion(x,n/2);
                        return val*val*x;
                    
                    
                }
        
        
        
    }
    
    
    double myPow(double x, int n) {
        
                    if(!n) return 1;
                        
            
                    if(n>0)
                        
                            return recursion(x,n);
                    
                    else{
                        
                            int val=abs(n); 
                            return 1/recursion(x,n);
                        
                        
                    }
        
                    return -1;
    }
};

File: 0050-powx-n/0050-powx-n.py
class Solution:
    
    
    def helper(self,x,n):
        
        if n==0:
            return 1
        
        val=self.helper(x,n//2)
        if n%2:
            return val*val*x
        else:
            return val*val
            
        
        
    
    
    
    def myPow(self, x: float, n: int) -> float:
        
        
        if x==1:
            return 1
        
        if n<0:
            return 1/self.helper(x,abs(n))
        return self.helper(x,n)
        
        